### ISSUE TITLE
Date/Time datatype

### ISSUE THREAD

**Issue by Nekomajin42** [josdejong/mathjs#384]
I'm following the evolution of mathjs towards v2 and I have a proposal about a new datatype.

It would be good to have the ability to do common operations like add and substract with time and date values. What do you think about it?

**Comment by josdejong** [comment]
Yes I like that idea. It's not useful for the JavaScript API (as there are exellent JS libraries for working with dates), but for the expression parser it can be a nice addition to be able to work with dates. We should think about what API we would like to offer then, and see if this can really work out nicely.

**Comment by FSMaxB** [comment]
Here are some thoughts about this:

I think it's important to allow different representations of dates, like `2015/5/31` or `31.05.2015` and so on with a global setting to switch between them (for the parser and the standard output format).

* There should also be functions to get the current time.
* Timezones?
* Represent points in time as well as time periods.

Operators:
* `time period * number`
* `time period / number`
* `time period + time period`
* `time period - time period`
* `point in time + time period`
* `point in time - time period`

As this is not changing anything existing this doesn't have to be in v2 necessarily because it can be introduced in a non-breaking matter. It can be introduced in later point versions, or am I wrong?

**Comment by josdejong** [comment]
Thanks for your input Max. I don't expect breaking changes here so there is no need to align this with the v2 release.

We indeed have to deal with Dates (point in time) and with periods (a duration or interval). The latter is already supported in very limited form: you can create units like `5 hour` or `30 minutes`. 

What we should prevent is writing our own `moment.js`. It would be great if we could enable using a library like `moment.js` in the expression parser with minimal effort. Working with dates is tricky, so I would really like to use an existing library here to prevent this taking on a life of its own. And it would be nice if we could use regular Units for time periods. We need to extend the time units then.

Just as an idea for the API:
```
30 minutes * 4   # already supported...

d = date('2015-06-01')   # returns an instance of moment.js
d + 4 hour               # we have to implement moment+Unit
                         # -> moment('2015-06-01').add(unit.value, 'seconds');
```

**Comment by tmtron** [comment]
If this is still on the list, you may consider to use the [TC39 temporal proposal](https://github.com/tc39/proposal-temporal)
* it uses immutable types
* currently (5.2021) it is a Stage 3 proposal, so there is already a working [polyfill](https://github.com/tc39/proposal-temporal/tree/main/polyfill) to test it

**Comment by m93a** [comment]
Using `Temporal` for date arithmetics sounds like a great idea to me. Personally, I'd love to look into it after #2212, but the ideas there might take a long time to get implemented, so if anybody wants to take on this earlier, feel free to :)

**Comment by josdejong** [comment]
Yes, still support for dates is still interesting. We'll have to think though an API in detail first I think, see how it can work out in the expression parser, and what to use underneath it (moment.js is outdated already). Good point @m93a about taking into account the ideas in #2212.
